rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,205 @@
+
+```yaml
+---
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+
+metadata:
+  name: eks-cluster-name
+  region: us-east-1
+
+identityProviders:
+  - name: TeleportProvider
+    type: oidc
+    issuerUrl: https://teleport.example.com
+    clientId: kubernetes
+    usernameClaim: teleport-eks-oidc-user
+    groupsClaim: teleport-eks-oidc-groups
+```
+
+To access the target EKS cluster in API mode, we will generate an access token using the associated OIDC provider.
+That token will grant access to the `system:masters` group, so the Helm chart installation will be able to set up
+all the required infrastructure for the Teleport kube agent, including service accounts.
+
+### Plan of implementation
+
+Implementation will be done in three main steps:
+
+1. Introducing a UI flow for the enrollment of selected EKS clusters using AWS CloudShell.
+2. Adding API-only enrollment for selected EKS clusters with the use of an associated OIDC identity provider.
+3. Adding the capability to set up automatic discovery and enrollment of EKS clusters.
+
+Each of these steps is functionally complete and allows building upon it to implement the next step.
+
+## Product Usage
+
+We will add a new PostHog event to track the usage of the Discover flow for EKS.
+
+```protobuf
+// UIDiscoverEKSClusterEnrollEvent is emitted when a user is finished with
+// the step that asks user to select from a list of EKS clusters.
+message UIDiscoverEKSClusterEnrollEvent {
+  DiscoverMetadata metadata = 1;
+  DiscoverResourceMetadata resource = 2;
+  DiscoverStepStatus status = 3;
+  int64 selected_resources_count = 4;
+}
+```
+
+## Security
+
+New permissions will be required for the AWS OIDC integration to perform the necessary tasks. It will need 
+to be able to list EKS clusters as well as associate an OIDC provider with them. An associated OIDC provider 
+can generate authentication tokens that give full control over the cluster, so a careful approach to their 
+security should be taken.
+
+Tokens will be signed by a dedicated OIDC Certificate Authority (which already exists and is used for 
+AWS integration communications), ensuring that tokens don't overlap with regular CAs access by the users, 
+like the User CA or JWTSigner CA. The audience field value `kubernetes` will also be unique to this type of token. 
+And also user/group claim field names, such as `teleport-eks-oidc-groups`, will specify them as used for this purpose only.
+
+End users will never have access to the process of generating these tokens; we will use them only to perform the 
+initial installation of the `teleport-kube-agent` Helm chart. The agent itself will be configured with the standard 
+service account permissions it requires.
+
+After the Helm chart installation succeeds, Teleport will no longer need to have an OIDC provider associated with 
+the enrolled EKS cluster. Clients can then remove the association or even impose a block on OIDC provider 
+associations for the cluster if they wish.
```

Comment:
Should teleport remove the association?
I'm sure Security folks would love that :sweat_smile:

Reply:
Yep, I think we can disassociate after we enrolled the cluster.